include/sparrow/arrow_array_schema_proxy.hpp

```diff
@@ -95,99 +139,176 @@ namespace sparrow
+        , m_data_ownership({.schema = ownership::owning, .array = ownership::owning})
     {
+        array = {};
+        schema = {};
         validate_array_and_schema();
-        initialize_children();
         initialize_buffers();
     }
 
-    inline
-    arrow_proxy::arrow_proxy(ArrowArray&& array, ArrowSchema* schema)
+    inline arrow_proxy::arrow_proxy(ArrowArray&& array, ArrowSchema* schema, ownership schema_ownership)
         : m_array(std::move(array))
+        , m_array_ref(std::get<1>(m_array))
         , m_schema(schema)
+        , m_schema_ref(*std::get<0>(m_schema))
+        , m_data_ownership({.schema = schema_ownership, .array = ownership::owning})
     {
+        array = {};
         SPARROW_ASSERT_TRUE(schema != nullptr);
         validate_array_and_schema();
-        initialize_children();
         initialize_buffers();
     }
 
-    inline
-    arrow_proxy::arrow_proxy(ArrowArray* array, ArrowSchema* schema)
+    inline arrow_proxy::arrow_proxy(
+        ArrowArray* array,
+        ownership array_ownership,
+        ArrowSchema* schema,
+        ownership schema_ownership
+    )
         : m_array(array)
+        , m_array_ref(*std::get<0>(m_array))
         , m_schema(schema)
+        , m_schema_ref(*std::get<0>(m_schema))
+        , m_data_ownership({.schema = schema_ownership, .array = array_ownership})
     {
         SPARROW_ASSERT_TRUE(array != nullptr);
         SPARROW_ASSERT_TRUE(schema != nullptr);
         validate_array_and_schema();
-        initialize_children();
         initialize_buffers();
     }
 
-    inline
-    arrow_proxy::~arrow_proxy()
+    arrow_proxy::arrow_proxy(const arrow_proxy& other)
+        : m_array(&other.m_array_ref)
+        , m_array_ref(*std::get<0>(m_array))
+        , m_schema(&other.m_schema_ref)
+        , m_schema_ref(*std::get<0>(m_schema))
+        , m_data_ownership(doesnt_own_arrow_data)
     {
-        if (m_array.index() == 1)
+        validate_array_and_schema();
+        initialize_buffers();
+    }
+
+    arrow_proxy& arrow_proxy::operator=(const arrow_proxy& other)
+    {
+        if (this == &other)
         {
-            ArrowArray& array = std::get<1>(m_array);
-            if (array.release != nullptr)
+            return *this;
+        }
+
+        m_array = &other.m_array_ref;
+        m_array_ref = *std::get<0>(m_array);
+        m_schema = &other.m_schema_ref;
+        m_schema_ref = *std::get<0>(m_schema);
+        m_data_ownership = doesnt_own_arrow_data;
+
+        validate_array_and_schema();
+        initialize_buffers();
+
+        return *this;
+    }
+
+    arrow_proxy::arrow_proxy(arrow_proxy&& other)
+        : m_array(std::move(other.m_array))
+        , m_array_ref(m_array.index() == 0 ? *std::get<0>(m_array) : std::get<1>(m_array))
+        , m_schema(std::move(other.m_schema))
+        , m_schema_ref(m_schema.index() == 0 ? *std::get<0>(m_schema) : std::get<1>(m_schema))
+        , m_data_ownership(other.m_data_ownership)
```

Comment:
note that you can move it too, it'll copy and if later the ownership data contain something else that's moveable you wont have to maintain that line

Reply:
Ok done

Reply:
removed finally: https://github.com/man-group/sparrow/pull/206#discussion_r1758423297